src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Android/OpenSslX509CertificateReader.cs

```diff
@@ -0,0 +1,572 @@
+                }
+
+                return context.Results;
+            }
+        }
+
+        internal static ArraySegment<byte> FindFirstExtension(SafeX509Handle cert, string oidValue)
+        {
+            return Interop.AndroidCrypto.X509FindExtensionData(cert, oidValue);
+        }
+
+        internal void SetPrivateKey(SafeEvpPKeyHandle privateKey)
+        {
+            _privateKey = privateKey;
+        }
+
+        internal SafeEvpPKeyHandle? PrivateKeyHandle
+        {
+            get { return _privateKey; }
+        }
+
+        public RSA? GetRSAPrivateKey()
+        {
+            if (_privateKey == null || _privateKey.IsInvalid)
+            {
+                return null;
+            }
+
+            return new RSAOpenSsl(_privateKey);
+        }
+
+        public DSA? GetDSAPrivateKey()
+        {
+            if (_privateKey == null || _privateKey.IsInvalid)
+            {
+                return null;
+            }
+
+            return new DSAOpenSsl(_privateKey);
+        }
+
+        public ECDsa GetECDsaPublicKey()
+        {
+            using (SafeEvpPKeyHandle publicKeyHandle = Interop.Crypto.GetX509EvpPublicKey(_cert))
+            {
+                Interop.Crypto.CheckValidOpenSslHandle(publicKeyHandle);
+
+                return new ECDsaOpenSsl(publicKeyHandle);
+            }
+        }
+
+        public ECDiffieHellman GetECDiffieHellmanPublicKey()
+        {
+            using (SafeEvpPKeyHandle publicKeyHandle = Interop.Crypto.GetX509EvpPublicKey(_cert))
+            {
+                Interop.Crypto.CheckValidOpenSslHandle(publicKeyHandle);
+
+                return new ECDiffieHellmanOpenSsl(publicKeyHandle);
+            }
+        }
+
+        public ECDsa? GetECDsaPrivateKey()
+        {
+            if (_privateKey == null || _privateKey.IsInvalid)
+            {
+                return null;
+            }
+
+            return new ECDsaOpenSsl(_privateKey);
+        }
+
+        public ECDiffieHellman? GetECDiffieHellmanPrivateKey()
+        {
+            if (_privateKey == null || _privateKey.IsInvalid)
+            {
+                return null;
+            }
+
+            return new ECDiffieHellmanOpenSsl(_privateKey);
+        }
+
+        private ICertificatePal CopyWithPrivateKey(SafeEvpPKeyHandle privateKey)
+        {
+            // This could be X509Duplicate for a full clone, but since OpenSSL certificates
+            // are functionally immutable (unlike Windows ones) an UpRef is sufficient.
+            SafeX509Handle certHandle = Interop.Crypto.X509UpRef(_cert);
+            OpenSslX509CertificateReader duplicate = new OpenSslX509CertificateReader(certHandle);
+
+            duplicate.SetPrivateKey(privateKey);
+            return duplicate;
+        }
+
+        public ICertificatePal CopyWithPrivateKey(DSA privateKey)
+        {
+            DSAOpenSsl? typedKey = privateKey as DSAOpenSsl;
```

Comment:
For all of the CopyWithPrivateKey overloads my feedback is "do we know what the Android implementation of these algorithms is going to look like?  Certainly in the end none of these will say OpenSsl, so it's unclear what "hey, that's a key object I gave you, I know how to work with that without going through the export routine" would really look like"

Reply:
This was the straight copy based on the original 'keep the shim APIs/communication like OpenSSL' desire (based on thinking that maybe handling the keys could be the same even if cert reading wasn't). But the more we have gotten into this, the less tenable that approach seems, so what exactly this will look like is not clear yet.

I think I will make these all throw NotImplemented for now, so that it is explicit.